core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -101,54 +100,62 @@ void release() {
         state.compareAndSet(State.FROZEN, State.RELEASED);
     }
 
-    ByteBuffer readMinKey() {
-        return entrySet.readKey(entrySet.getHeadNextIndex());
+    boolean readMinKey(EntrySet.KeyBuffer key) {
+        return entrySet.readKey(key, entrySet.getHeadNextIndex());
     }
 
-    ByteBuffer readMaxKey() {
+    boolean readMaxKey(EntrySet.KeyBuffer key) {
         int maxEntry = getLastItemEntryIndex();
-        return entrySet.readKey(maxEntry);
+        return entrySet.readKey(key, maxEntry);
     }
 
     /**
      * look up key
      *
+     * @param ctx a context object to update with the results
      * @param key the key to look up
-     * @return a LookUp object describing the condition of the value associated with {@code key}.
-     * If {@code lookup == null}, there is no entry with that key in this chunk.
-     * If {@code lookup.valueReference == INVALID_VALUE}, it means that there is an entry with that key (can be
+     *
+     * The ctx will describe the condition of the value associated with {@code key}.
+     * If {@code (ctx.isKeyValid() && ctx.isValueValid()) == True}, the key and value was found.
+     * If {@code ctx.isKeyValid() == False}, there is no entry with that key in this chunk.
+     * Otherwise, if {@code lookup.isValueValid() == False}, it means that there is an entry with that key (can be
      * reused in case
      * of {@code put}, but there is no value attached to this key (it can happen if this entry is in the midst of
      * being inserted, or some thread removed this key and no rebalanace occurred leaving the entry in the chunk).
-     * It implies that {@code lookup.valueSlice = null}.
-     * If {@code lookup.valueSlice == null && lookup.valueReference != INVALID_VALUE} it means that the value is
-     * marked off-heap as deleted, but the connection between the entry and the value was not unlinked yet.
+     * It means that the value is marked off-heap as deleted, but the connection between the entry and the value was not unlinked yet.
+     * {@code ctx.valueState} will indicate the reason.
      */
-    EntrySet.LookUp lookUp(K key) {
+    void lookUp(ThreadContext ctx, K key) {
         // binary search sorted part of key array to quickly find node to start search at
         // it finds previous-to-key
-        int curr = binaryFind(key);
+        int curr = binaryFind(ctx.key, key);
         curr = (curr == NONE_NEXT) ? entrySet.getHeadNextIndex() : entrySet.getNextEntryIndex(curr);
-        int cmp;
-        // iterate until end of list (or key is found)
 
+        // iterate until end of list (or key is found)
         while (curr != NONE_NEXT) {
             // compare current item's key to searched key
-            cmp = comparator.compareKeyAndSerializedKey(key, entrySet.readKey(curr));
+            boolean isValid = entrySet.keyLookup(ctx, curr);
```

Comment:
I am bothered with the additional amount of function calls added here relative to the previous version of the code... Can we do something about it?